src/core/eme/create_session.ts

```diff
@@ -0,0 +1,198 @@
+  return Observable.defer(() => {
+    const {
+      keySystemOptions,
+      mediaKeySystemAccess,
+      sessionsStore,
+      sessionStorage,
+    } = mediaKeysInfos;
+
+    const mksConfig = mediaKeySystemAccess.getConfiguration();
+    const sessionTypes = mksConfig.sessionTypes;
+    const hasPersistence = (
+      sessionTypes && arrayIncludes(sessionTypes, "persistent-license")
+    );
+
+    const sessionType : MediaKeySessionType =
+      hasPersistence &&
+      sessionStorage &&
+      keySystemOptions.persistentLicense ?
+      "persistent-license" : "temporary";
+
+    log.debug(`eme: create a new ${sessionType} session`);
+
+    const session = sessionsStore.createSession(initData, initDataType, sessionType);
+
+    // Re-check for Dumb typescript
+    if (!hasPersistence || !sessionStorage || !keySystemOptions.persistentLicense) {
+      if (__DEV__) {
+        assert(sessionType === "temporary");
+      }
+      return Observable.of({
+        type: "created-session" as "created-session",
+        value: { mediaKeySession: session, sessionType },
+      });
+    }
+    if (__DEV__) {
+      assert(sessionType === "persistent-license");
+    }
+
+    const storedEntry = sessionStorage.get(initData, initDataType);
+    if (!storedEntry) {
+      return Observable.of({
+        type: "created-session" as "created-session",
+        value: { mediaKeySession: session, sessionType },
+      });
+    }
+
+    return loadPersistentSession(storedEntry.sessionId, session)
+      .mergeMap((hasLoadedSession) : Observable<ICreateSessionEvent> => {
+        if (!hasLoadedSession) {
+          log.warn("eme: no data stored for the loaded session");
+          sessionStorage.delete(initData, initDataType);
+          return Observable.of({
+            type: "created-session" as "created-session",
+            value: { mediaKeySession: session, sessionType },
+          });
+        }
+
+        if (hasLoadedSession && isSessionUsable(session)) {
+          sessionStorage.add(initData, initDataType, session);
+          return Observable.of({
+            type: "loaded-persistent-session" as "loaded-persistent-session",
+            value: { mediaKeySession: session, sessionType },
+          });
+        }
+
+        // Unusable persistent session: recreate a new session from scratch.
+        return recreateSessionForInitData(
+          session, sessionType, sessionsStore, initData, initDataType)
+            .map((newSession) => {
+              return {
+                type: "created-session" as "created-session",
+                value: { mediaKeySession: newSession, sessionType },
+              };
+            });
+      })
+      .catch(() : Observable<INewSessionCreatedEvent> => {
+        // Failure to load persistent session: recreate a new session from scratch.
+        return recreateSessionForInitData(
+          session, sessionType, sessionsStore, initData, initDataType)
+            .map((newSession) => {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Update this function so that its implementation is not identical to the one on line 179. ([typescript:S4144](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4144))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWNEQ0wcneya_4cHG_Dj&open=AWNEQ0wcneya_4cHG_Dj)

[ISSUE_KEY]: AWNEQ0wcneya_4cHG_Dj